File: algorithms/js/validAnagram/validAnagram.js
/*
 * @Source: https://leetcode-cn.com/problems/valid-anagram/
 * @Author: Tian Zhuo
 * @Date: 2019-01-14 22:46:43
 */
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
    const sort = str => [...str]
        .sort((a, b) => a.charCodeAt() - b.charCodeAt())
        .join('')
    return sort(s) === sort(t)
};

var isAnagram = function (s, t) {
    let len = s.length;
    for (let i = 0; i < len; i++) {
        if (s === t || s.length !== t.length) break;
        t = t.replace(new RegExp(s[0], 'g'), '');
        s = s.replace(new RegExp(s[0], 'g'), '');
    }
    return s === t;
};

File: src/main/java/validAnagram/ValidAnagram.java
/*
 * @Source: https://leetcode-cn.com/problems/valid-anagram/
 * @Author: Tian Zhuo
 * @Date: 2019-01-14 21:36:23
 */
package validAnagram;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class ValidAnagram {
    // 20 ms
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()) return false;
        Map<Character, Integer> map = new HashMap<>();
        for (char ch : s.toCharArray()) {
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }
        for (char ch : t.toCharArray()) {
            Integer c = map.get(ch);
            if(c == null){
                return false;
            }else if(c > 1){
                map.put(ch, c - 1);
            }else{
                map.remove(ch);
            }
        }
        return map.isEmpty();
    }

    // 3 ms
    public boolean isAnagram1(String s, String t) {
        char[] sChars = s.toCharArray();
        char[] tChars = t.toCharArray();
        Arrays.sort(sChars);
        Arrays.sort(tChars);
        return String.valueOf(sChars).equals(String.valueOf(tChars));
    }

    // 6 ms
    public boolean isAnagram2(String s, String t) {
        int[] sCounts = new int[26];
        int[] tCounts = new int[26];
        for (char ch : s.toCharArray()) {
            sCounts[ch - 'a']++;
        }
        for (char ch : t.toCharArray()) {
            tCounts[ch - 'a']++;
        }
        for (int i = 0; i < 26; i++) {
            if (sCounts[i] != tCounts[i]) {
                return false;
            }
        }
        return true;
    }
}